### ISSUE TITLE
MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups

### ISSUE THREAD

**Issue by hasancelik** [hazelcast/hazelcast#7959]
```
java.lang.AssertionError: expected:<10000> but was:<9840>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:645)
    at org.junit.Assert.assertEquals(Assert.java:631)
    at com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups(MapLoaderFailoverTest.java:164)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/881/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups/

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK1.7/com.hazelcast$hazelcast/454/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups/

**Comment by hasancelik** [comment]
suppressed by https://github.com/hazelcast/hazelcast/commit/2543fd9c08fdd81f20267190d798bcf56af9da51

**Comment by tombujok** [comment]
Managed to reproduce with Repeat(1000) on
```
java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode)
```

**Comment by tombujok** [comment]
OK, I found the root cause.

The problem is that sometimes after the migration the migrated partition gets replicaIndex == 0 (according to the migration log), but the `partitionService.isPartitionOwner(partitionId);` call just after this returns false. 

Here's a log snippet:
```
BEFORE_MIGRATION on 2 old=0 new=-1 SOURCE
BEFORE_MIGRATION on 2 old=2 new=0 DESTINATION
(PartitionContainer.recordStoreConstructor) CREATING RECORD_STORE partitionId on 2 this address = Address[127.0.0.1]:5087
(recordStore.startLoading)      startInitialLoad on 2
(MapKeyLoader.startInitialLoad) isPartitionOwner=false isMapNamePartition=false isMapNamePartitionFirstReplica=false hasBackup=false ROLE=NONE partitionId=2 address=Address[127.0.0.1]:5087
(MapKeyLoader.startInitialLoad) NO ROLE on 2
```

You can see the BEFORE_MIGRATION event that gives the 2nd partition replicaIndex==0 on DESTINATION, but the call in maploader `partitionService.isPartitionOwner(partitionId);` returns false, and that partition gets no ROLE in map-loader (since map-loader thinks it's a backup partition). 

That call should return true - maybe there's a race there or sth.

It's easy to reproduce this problem, just run it with @Repeat(100) - it fails within a minute.
The easiest way to debug this behavior is to add a log message in MapKeyLoader.startInitialLoad
```
public Future startInitialLoad(MapStoreContext mapStoreContext, int partitionId) {

        this.partitionId = partitionId;
        this.mapNamePartition = partitionService.getPartitionId(toData.apply(mapName));
        Role newRole = calculateRole();

        role.nextOrStay(newRole);
        state.next(State.LOADING);

        switch (newRole) {
            case SENDER:
                return sendKeys(mapStoreContext, false);
            case SENDER_BACKUP:
                return triggerLoading();
            case RECEIVER:
                return triggerLoading();
            default:
                System.err.println("!!!!!!!!!!!!!!  NO ROLE on " + partitionId);
                return loadFinished;
        }
    }
```

If you see NO ROLE it means that the partitionService returned the wrong result. It happens only in tests run that fail

**Comment by tombujok** [comment]
It may be caused by the fact that you first create DefaultRecordStore on migration and then update the PartitionTable. MapLoader checks if the DefaultRecordStore is the owner of that partition in the startInitialLoad function that's called in the `PartitionContainer.recordStoreConstructor`

**Comment by tombujok** [comment]
This call is responsible for the creation of the DefaultRecordStore on migration
https://github.com/tombujok/hazelcast/blob/02433cdce1e6624f90be5bbb5f28e0b05225deb6/hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationOperation.java#L146

but there's also a second one here (I'm not sure when it's called)
https://github.com/tombujok/hazelcast/blob/02433cdce1e6624f90be5bbb5f28e0b05225deb6/hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationOperation.java#L240

Basically, the root cause is that this call cannot be executed before the migration.commit() call
https://github.com/tombujok/hazelcast/blob/97f34870ca2d60d0aac2d7f3689160061442444e/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java#L48-L48

For now it's not the case, thus the map-loader cannot calculate its role properly.

**Comment by tombujok** [comment]
The threading works as follows:
the top-level method that calls `isPartitionOwner` is this one:
https://github.com/tombujok/hazelcast/blob/97f34870ca2d60d0aac2d7f3689160061442444e/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java#L48-L48

It is called from the thread that creates the record store (calls the `startLoading` method)

**Comment by mmedenjak** [comment]
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3199

**Comment by vbekiaris** [comment]
The test failed with "expected:<2> but was:<3>" at "testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups". This means that the map was properly loaded but initial key loading was invoked three times instead of 2. In the logs I don't see any explanation why initial key loading was retried twice, but this doesn't look like a real issue to me. Probably we should change the test to assert that key loading is tried 2 or more times.

**Comment by sancar** [comment]
@vbekiaris I new failure
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/5392/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups/
```
java.lang.AssertionError: expected:<2> but was:<3>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups(MapLoaderFailoverTest.java:190)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```
